### ISSUE TITLE
[feature request] Support of decoupled registration of instrumentation with support of asynchronous code execution

### ISSUE THREAD

**Issue by plachor** [open-telemetry/opentelemetry-dotnet#5876]
### Package

None

### Is your feature request related to a problem?

Hi, I'm exploring best options to setup telemetry in my project and it seems that options of registration of Tracing instrumentation is quite limited. It looks like there is:
 * no async support;
 * there is builder which must be invoked in one place and aggregate all configuration there at phase of IServiceCollection construction;

My project is set of core libraries that help build micrsoservices and is divided into small components (projects) like EF, Elasticsearch, Redis, Kafka, S3, ASPNet and similar. There is shared project that delivers abstractions however concrete implementations are implemented and registered within own projects and can be used on demand by end developer who will create his own microservice using that. 

It would make more sense for me to add Npgsql instrumentation near bootstrap code of Npqsql driver and put all customization next to that code. So that related features are coupled together and registered only when needed.

Further more some of my services and connection parameters are obtained asynchronously for instance from Secret Storage yet during registration I need them all upfront known in order to enable telemetry like for instance for Redis where I must provide IConnectionMultiplexer in synchronous way. @

### What is the expected behavior?

I would love more flexible option setup instrumentation similar to what IOptions enable. Like for instance on Redis example:

```csharp
public static IServiceCollection AddRedisInstrumentation(
        this IServiceCollection sc, string? traceSetupName
        Action<IServiceProvider, StackExchangeRedisInstrumentationOptions> configure)
    {
          ....
    }
    
    public static IServiceCollection AddRedisInstrumentation<TConfigureInstrumentationService>(this IServiceCollection sc, string? traceSetupName)
         where TConfigureInstrumentationService : IConfigureInstrumentationService<StackExchangeRedisInstrumentationOptions>
    {
          ....
    }
    
    public interface IConfigureInstrumentationService<TInstrumentationOptions> where TInstrumentationOptions : IInstrumentationOptions
   {
         ValueTask ConfigureAsync(TInstrumentationOptions options);
   }
```
It would allow to decouple registration and add asynchronous configuration 

### Which alternative solutions or features have you considered?

Looking still for them I guess I will add Otel at end and scan IServiceConfigruation for my custom registrations and execute them in foreach at the end once everything is registered before DI construction but it is not best to do I believe.

**Comment by CodeBlanch** [comment]
> there is builder which must be invoked in one place and aggregate all configuration there at phase of IServiceCollection construction;

False! 😄

See: https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/extending-the-sdk#iservicecollection-extension-methods

There are these what I call "detached" extensions:

* `IServiceCollection.ConfigureOpenTelemetryLoggerProvider`
* `IServiceCollection.ConfigureOpenTelemetryMeterProvider`
* `IServiceCollection.ConfigureOpenTelemetryTracerProvider`

If you have all your libraries call those methods then in your main host/app all you need is a single call to turn on the SDK like `services.AddOpenTelemetry().WithLogging().WithMetrics().WithTracing()`.

> no async support;

I need to understand more about this use case. In the years I've been involved in this project I can't recall a single user asking for this so it doesn't seem to be a blocker. And don't say it is because you are using secrets. I'm sure there are lots of users successfully using secrets and OTel 😄 Could you maybe provide a small sample of what you _have_ to do and what you would _like_ to do and we can go from there?

**Comment by plachor** [comment]
Hi @CodeBlanch, thanks for you prompt response - I've tried to respond on Friday however had issue with conditional service registration (which seems to be by design as explained in XML comment): https://github.com/open-telemetry/opentelemetry-dotnet/blob/5106e3082f92bcd1cc0a3a259198f116a4a51fb1/src/OpenTelemetry.Api.ProviderBuilderExtensions/Trace/OpenTelemetryDependencyInjectionTracingServiceCollectionExtensions.cs#L63-L70 

In you sample you show exact usage I would like to achieve: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/extending-the-sdk/README.md?plain=1#L583-L590

I would like to be able for instance to add Redis instrumentation but only in case telemetry is enabled (so all profiling stuff is not gathered in case it will not be used). I can still achieve that as I can access `IConfiguration` already registered in `IServiceCollection` before service provider is constructed yet it seems like the overload with service provider is quite limited for now (will work only with basic instrumentations).

When it comes to async support I found that in my case (again Redis) I can post-configure connection multiplex later on using https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/c32fea375a52344ee1ebbf219675decb2699c9bd/src/OpenTelemetry.Instrumentation.StackExchangeRedis/StackExchangeRedisInstrumentation.cs#L39 so it seems I'm also suited here.

And to explain why: in my case in order to construct connection multiplexer I need to execute first several asynchronous HTTP calls. That's why I needed `async` support.  

Thnx for you assitance.
